pkg/vmtest/integration.go

```diff
@@ -236,168 +214,133 @@ func QEMUTest(t *testing.T, o *Options) (*qemu.VM, func()) {
 // QEMU VM, or an error.
 func QEMU(o *Options) (*qemu.Options, string, error) {
 	if len(o.Name) == 0 {
 		o.Name = callerName(2)
 	}
 
-	if o.Env == nil {
-		env := golang.Default()
-		o.Env = &env
-		o.Env.CgoEnabled = false
-		env.GOARCH = TestArch()
-	}
+	if len(o.QEMUOpts.Initramfs) == 0 {
+		if !o.DontSetEnv {
+			env := golang.Default()
+			env.CgoEnabled = false
+			env.GOARCH = TestArch()
+			o.BuildOpts.Env = env
+		}
 
-	if len(o.LogFile) == 0 {
-		// Create file for serial logs.
-		if err := os.MkdirAll(logDir, 0755); err != nil {
-			return nil, "", fmt.Errorf("could not create serial log directory: %v", err)
+		var cmds []string
+		if len(o.BuildOpts.Commands) == 0 {
+			cmds = append(cmds, "github.com/u-root/u-root/cmds/*")
 		}
+		// Create a uinit from the commands given.
+		if len(o.Uinit) > 0 {
+			urootPkg, err := o.BuildOpts.Env.Package("github.com/u-root/u-root/integration")
+			if err != nil {
+				return nil, "", err
+			}
+			testDir := filepath.Join(urootPkg.Dir, "testcmd")
 
-		o.LogFile = filepath.Join(logDir, fmt.Sprintf("%s.log", o.Name))
-	}
+			dirpath, err := ioutil.TempDir(testDir, "uinit-")
+			if err != nil {
+				return nil, "", err
+			}
+			defer os.RemoveAll(dirpath)
 
-	var cmds []string
-	if len(o.Cmds) == 0 {
-		cmds = append(cmds, "github.com/u-root/u-root/cmds/*")
-	} else {
-		cmds = append(cmds, o.Cmds...)
-	}
-	// Create a uinit from the commands given.
-	if len(o.Uinit) > 0 {
-		urootPkg, err := o.Env.Package("github.com/u-root/u-root/integration")
-		if err != nil {
-			return nil, "", err
-		}
-		testDir := filepath.Join(urootPkg.Dir, "testcmd")
+			if err := os.MkdirAll(filepath.Join(dirpath, "uinit"), 0755); err != nil {
+				return nil, "", err
+			}
 
-		dirpath, err := ioutil.TempDir(testDir, "uinit-")
-		if err != nil {
-			return nil, "", err
+			var realUinit [][]string
+			for _, cmd := range o.Uinit {
+				realUinit = append(realUinit, fields(cmd))
+			}
+
+			if err := ioutil.WriteFile(
+				filepath.Join(dirpath, "uinit", "uinit.go"),
+				[]byte(fmt.Sprintf(template, realUinit)),
+				0755); err != nil {
+				return nil, "", err
+			}
+			cmds = append(cmds, path.Join("github.com/u-root/u-root/integration/testcmd", filepath.Base(dirpath), "uinit"))
+		}
+		// Add our commands to the build opts.
+		if len(cmds) > 0 {
+			o.BuildOpts.AddBusyBoxCommands(cmds...)
 		}
-		defer os.RemoveAll(dirpath)
 
-		if err := os.MkdirAll(filepath.Join(dirpath, "uinit"), 0755); err != nil {
-			return nil, "", err
+		// Create or reuse a temporary directory.
+		if len(o.BuildOpts.TempDir) == 0 {
+			tmpDir, err := ioutil.TempDir("", "uroot-integration")
+			if err != nil {
+				return nil, "", err
+			}
+			o.BuildOpts.TempDir = tmpDir
+		}
+		if o.BuildOpts.BaseArchive == nil {
+			o.BuildOpts.BaseArchive = uroot.DefaultRamfs.Reader()
+		}
+		if len(o.BuildOpts.InitCmd) == 0 {
+			o.BuildOpts.InitCmd = "init"
+		}
+		if len(o.BuildOpts.DefaultShell) == 0 {
+			o.BuildOpts.DefaultShell = "elvish"
```

Comment:
not /bin/sh? Just wondering. I'm not quite able to connect all the dots, so I'm not sure if the /bin/sh link gets made in this process.

Reply:
This is the option that determined what /bin/sh and /bin/defaultsh link to